pigs.1

." Process this file with
." groff -man -Tascii foo.1
."
.TH pigs 1 2012-2020 Linux "pigpio archive"
.SH NAME
pigs - command line socket access to the pigpio daemon.

/dev/pigpio - command line pipe access to the pigpio daemon.

.SH SYNOPSIS

.B sudo pigpiod

then

.B pigs {command}+

or

.B "echo {command}+ >/dev/pigpio"

.SH DESCRIPTION

.ad l

.nh


.br
The socket and pipe interfaces allow control of the Pi's GPIO by
passing messages to the running pigpio library.

.br
The normal way to start the pigpio library would be as a daemon during boot.

.br

.EX
sudo pigpiod
.br

.EE

.br
.SS Features
.br
o hardware timed PWM on any of GPIO 0-31

.br
o hardware timed servo pulses on any of GPIO 0-31

.br
o reading/writing all of the GPIO in a bank as one operation

.br
o individually setting GPIO modes, reading and writing

.br
o notifications when any of GPIO 0-31 change state

.br
o the construction of output waveforms with microsecond timing

.br
o I2C, SPI, and serial link wrappers

.br
o creating and running scripts on the pigpio daemon

.br
.SS GPIO
.br
ALL GPIO are identified by their Broadcom number.

.br
.SS Usage
.br
pigs is a program and internally uses the socket interface to pigpio
whereas /dev/pigpio uses the pipe interface.

.br
pigs and the pipe interface share the same commands and are invoked in
a similar fashion from the command line.

.br
The pigpio library must be running, either by running a program linked
with the library or starting the pigpio daemon (sudo pigpiod).

.br
pigs {command}+

.br
echo "{command}+" >/dev/pigpio

.br
pigs will show the result of the command on screen.

.br
The pigs process returns an exit status (which can be displayed with
the command echo $?).

.br

.EX
PIGS_OK            0
.br
PIGS_CONNECT_ERR 255
.br
PIGS_OPTION_ERR  254
.br
PIGS_SCRIPT_ERR  253
.br
.br
.br

.EE

.br
The results of /dev/pigpio commands need to be read from /dev/pigout,
e.g. cat /dev/pigout (try cat /dev/pigout& so that all subsequent
results are shown on screen).

.br
In both cases if an error was detected a message will have been written
to /dev/pigerr (try cat /dev/pigerr&).  This is likely to be more
informative than the message returned by pigs or the error code
returned by the pipe interface.

.br
Several commands may be entered on a line.  If present PROC and PARSE must
be the last command on a line.

.br
E.g.

.br

.EX
pigs w 22 1 mils 1000 w 22 0
.br

.EE

.br
is equivalent to

.br

.EX
pigs w 22 1
.br
pigs mils 1000
.br
pigs w 22 0
.br

.EE

.br
and

.br

.EX
echo "m 4 w w 4 0 mils 250 m 4 r r 4" >/dev/pigpio
.br

.EE

.br
is equivalent to

.br

.EX
echo "m 4 w"    >/dev/pigpio
.br
echo "w 4 0"    >/dev/pigpio
.br
echo "mils 250" >/dev/pigpio
.br
echo "m 4 r"    >/dev/pigpio
.br
echo "r 4"      >/dev/pigpio
.br

.EE

.br
.SS Notes
.br
The examples from now on will show the pigs interface but the same
commands will also work on the pipe interface.

.br
pigs does not show the status of successful commands unless the
command itself returns data.  The status (0) will be returned to
pigs but will be discarded.

.br
The status/data of each command sent to the pipe interface should
be read from /dev/pigout.

.br
When a command takes a number as a parameter it may be entered as hex
(precede by 0x), octal (precede by 0), or decimal.

.br
E.g. 23 is 23 decimal, 0x100 is 256 decimal, 070 is 56 decimal.

.br
Some commands can return a variable number of data bytes.  By
default this data is displayed as decimal.  The pigs -a option
can be used to force the display as ASCII and the pigs -x
option can be used to force the display as hex.

.br
E.g. assuming the transmitted serial data is the letters ABCDEONM

.br

.EX
$ pigs slr 4 100
.br
8 65 66 67 68 69 79 78 77
.br

.br
$ pigs -a slr 4 100
.br
8 ABCDEONM
.br

.br
$ pigs -x slr 4 100
.br
8 41 42 43 44 45 4f 4e 4d
.br

.EE

.br

.SH OVERVIEW
.SS BASIC
.B M/MODES g m 
Set GPIO mode 
.P
.B MG/MODEG g 
Get GPIO mode 
.P
.B PUD g p 
Set GPIO pull up/down 
.P
.B R/READ g 
Read GPIO level  
.P
.B W/WRITE g L 
Write GPIO level 
.P
.SS PWM (overrides servo commands on same GPIO)
.B P/PWM u v 
Set GPIO PWM value      
.P
.B PFS u v 
Set GPIO PWM frequency  
.P
.B PRS u v 
Set GPIO PWM range      
.P
.B GDC u 
Get GPIO PWM dutycycle  
.P
.B PFG u 
Get GPIO PWM frequency  
.P
.B PRG u 
Get GPIO PWM range      
.P
.B PRRG u 
Get GPIO PWM real range 
.P
.SS Servo (overrides PWM commands on same GPIO)
.B S/SERVO u v 
Set GPIO servo pulsewidth 
.P
.B GPW u 
Get GPIO servo pulsewidth 
.P
.SS INTERMEDIATE
.B TRIG u pl L 
Send a trigger pulse 
.P
.B WDOG u v 
Set GPIO watchdog    
.P
.B BR1  
Read bank 1 GPIO               
.P
.B BR2  
Read bank 2 GPIO               
.P
.B BC1 bits 
Clear specified GPIO in bank 1 
.P
.B BC2 bits 
Clear specified GPIO in bank 2 
.P
.B BS1 bits 
Set specified GPIO in bank 1   
.P
.B BS2 bits 
Set specified GPIO in bank 2   
.P
.SS ADVANCED
.B NO  
Request a notification 
.P
.B NC h 
Close notification     
.P
.B NB h bits 
Start notification     
.P
.B NP h 
Pause notification     
.P
.B HC g cf 
Set hardware clock frequency 
.P
.B HP g pf pdc 
Set hardware PWM frequency and dutycycle 
.P
.B FG u stdy 
Set a glitch filter on a GPIO  
.P
.B FN u stdy actv 
Set a noise filter on a GPIO   
.P
.B PADS pad padma 
Set pad drive strength 
.P
.B PADG pad 
Get pad drive strength 
.P
.B SHELL name str 
Execute a shell command 
.P
.SS Custom
.B CF1 uvs 
Custom function 1 
.P
.B CF2 uvs 
Custom function 2 
.P
.SS Events
.B EVM h bits 
Set events to monitor 
.P
.B EVT event 
Trigger event         
.P
.SS Scripts
.B PROC t 
Store script                     
.P
.B PROCR sid pars 
Run script                       
.P
.B PROCU sid pars 
Set script parameters            
.P
.B PROCP sid 
Get script status and parameters 
.P
.B PROCS sid 
Stop script                      
.P
.B PROCD sid 
Delete script                    
.P
.B PARSE t 
Validate script                  
.P
.SS I2C
.B I2CO ib id if 
Open I2C bus and device with flags 
.P
.B I2CC h 
Close I2C handle                   
.P
.B I2CWQ h bit 
smb Write Quick: write bit 
.P
.B I2CRS h 
smb Read Byte: read byte   
.P
.B I2CWS h bv 
smb Write Byte: write byte 
.P
.B I2CRB h r 
smb Read Byte Data: read byte from register 
.P
.B I2CWB h r bv 
smb Write Byte Data: write byte to register 
.P
.B I2CRW h r 
smb Read Word Data: read word from register 
.P
.B I2CWW h r wv 
smb Write Word Data: write word to register 
.P
.B I2CRK h r 
smb Read Block Data: read data from register 
.P
.B I2CWK h r bvs 
smb Write Block Data: write data to register 
.P
.B I2CWI h r bvs 
smb Write I2C Block Data                          
.P
.B I2CRI h r num 
smb Read I2C Block Data: read bytes from register 
.P
.B I2CRD h num 
i2c Read device  
.P
.B I2CWD h bvs 
i2c Write device 
.P
.B I2CPC h r wv 
smb Process Call: exchange register with word             
.P
.B I2CPK h r bvs 
smb Block Process Call: exchange data bytes with register 
.P
.B I2CZ h bvs 
Performs multiple I2C transactions 
.P
.SS I2C BIT BANG
.B BI2CO sda scl b 
Open bit bang I2C  
.P
.B BI2CC sda 
Close bit bang I2C 
.P
.B BI2CZ sda bvs 
I2C bit bang multiple transactions 
.P
.SS I2C/SPI SLAVE
.B BSCX bctl bvs 
BSC I2C/SPI transfer 
.P
.SS SERIAL
.B SERO dev b sef 
Open serial device dev at baud b with flags 
.P
.B SERC h 
Close serial handle                         
.P
.B SERRB  
Read byte from serial handle  
.P
.B SERWB h bv 
Write byte to serial handle   
.P
.B SERR h num 
Read bytes from serial handle 
.P
.B SERW h bvs 
Write bytes to serial handle  
.P
.B SERDA h 
Check for serial data ready to read 
.P
.SS SERIAL BIT BANG (read only)
.B SLRO u b db 
Open GPIO for bit bang serial data     
.P
.B SLRC u 
Close GPIO for bit bang serial data    
.P
.B SLRI u v 
Sets bit bang serial data logic levels 
.P
.B SLR u num 
Read bit bang serial data from GPIO    
.P
.SS SPI
.B SPIO c b spf 
SPI open channel at baud b with flags 
.P
.B SPIC h 
SPI close handle                    
.P
.B SPIR h num 
SPI read bytes from handle          
.P
.B SPIW h bvs 
SPI write bytes to handle           
.P
.B SPIX h bvs 
SPI transfer bytes to handle        
.P
.SS SPI BIT BANG
.B BSPIO cs miso mosi sclk b spf 
Open bit bang SPI      
.P
.B BSPIC cs 
Close bit bang SPI    
.P
.B BSPIX cs bvs 
SPI bit bang transfer 
.P
.SS FILES
.B FO file mode 
Open a file in mode            
.P
.B FC h 
Close file handle              
.P
.B FR h num 
Read bytes from file handle    
.P
.B FW h bvs 
Write bytes to file handle     
.P
.B FS h num from 
Seek to file handle position   
.P
.B FL pat num 
List files which match pattern 
.P
.SS WAVES
.B WVCLR  
Clear all waveforms 
.P
.B WVNEW  
Initialise a new waveform    
.P
.B WVAG trips 
Add generic pulses to waveform 
.P
.B WVAS u b db sb o bvs 
Add serial data to waveform    
.P
.B WVCRE  
Create a waveform   
.P
.B WVCAP percent 
Create a waveform of fixed size 
.P
.B WVDEL wid 
Delete selected waveform 
.P
.B WVTX wid 
Transmits waveform once       
.P
.B WVTXM wid wmde 
Transmits waveform using mode 
.P
.B WVTXR wid 
Transmits waveform repeatedly 
.P
.B WVCHA bvs 
Transmits a chain of waveforms 
.P
.B WVTAT  
Returns the current transmitting waveform 
.P
.B WVBSY  
Check if waveform is being transmitted 
.P
.B WVHLT  
Stop waveform 
.P
.B WVSC ws 
Get waveform DMA CB stats    
.P
.B WVSM ws 
Get waveform time stats      
.P
.B WVSP ws 
Get waveform pulse stats     
.P
.SS UTILITIES
.B H/HELP  
Display command help        
.P
.B HWVER  
Get hardware version
.P
.B SERNM
Get hardware serial number
.P
.B MICS v 
Microseconds delay          
.P
.B MILS v 
Milliseconds delay          
.P
.B PIGPV  
Get pigpio library version  
.P
.B T/TICK  
Get current tick            
.P
.SS CONFIGURATION
.B CGI  
Configuration get internals 
.P
.B CSI v 
Configuration set internals 
.P

.SH COMMANDS

.br

.IP "\fBBC1 bits\fP - Clear specified GPIO in bank 1"
.IP "" 4
This command clears (sets low) the GPIO specified by \fBbits\fP in bank 1.
Bank 1 consists of GPIO 0-31.

.br
Upon success nothing is returned.  On error a negative status code
will be returned.

.br

\fBExample\fP
.br

.EX
$ pigs bc1 0x400010 # clear GPIO 4 (1<<4) and 22 (1<<22)
.br

.br
$ pigs bc1 32 # clear GPIO 5 (1<<5)
.br
-42
.br
ERROR: no permission to update one or more GPIO
.br

.EE

.br

.IP "\fBBC2 bits\fP - Clear specified GPIO in bank 2"
.IP "" 4
This command clears (sets low) the GPIO specified by \fBbits\fP in bank 2.
Bank 2 consists of GPIO 32-53.

.br
Upon success nothing is returned.  On error a negative status code
will be returned.

.br

\fBExample\fP
.br

.EX
$ pigs bc2 0x8000 # clear GPIO 47 (activity LED on A+/B+/Pi2/Pi3)
.br

.br
$ pigs bc2 1 # clear GPIO 32 (first in bank 2)
.br
-42
.br
ERROR: no permission to update one or more GPIO
.br

.EE

.br

.IP "\fBBI2CC sda\fP - Close bit bang I2C"
.IP "" 4
This command signals that bit banging I2C on \fBsda\fP (and \fBscl\fP) is no
longer required.

.br

\fBExample\fP
.br

.EX
$ pigs bi2cc 5
.br

.EE

.br

.IP "\fBBI2CO sda scl b\fP - Open bit bang I2C"
.IP "" 4
This command signals that GPIO \fBsda\fP and \fBscl\fP are to be used
for bit banging I2C at \fBb\fP baud.

.br
Bit banging I2C allows for certain operations which are not possible
with the standard I2C driver.

.br
o baud rates as low as 50
.br
o repeated starts
.br
o clock stretching
.br
o I2C on any pair of spare GPIO

.br
The baud rate may be between 50 and 500000 bits per second.

.br
The GPIO used for SDA and SCL must have pull-ups to 3V3 connected.  As
a guide the hardware pull-ups on pins 3 and 5 are 1k8 in value.

.br

.IP "\fBBI2CZ sda bvs\fP - I2C bit bang multiple transactions"
.IP "" 4
This function executes a sequence of bit banged I2C operations.  The
operations to be performed are specified by the contents of \fBbvs\fP
which contains the concatenated command codes and associated data.

.br
The following command codes are supported:

.br

.EX
Name      Cmd & Data     Meaning
End       0              No more commands
Escape    1              Next P is two bytes
Start     2              Start condition
Stop      3              Stop condition
Address   4 P            Set I2C address to P
Flags     5 lsb msb      Set I2C flags to lsb + (msb << 8)
Read      6 P            Read P bytes of data
Write     7 P ...        Write P bytes of data

.EE

.br
The address, read, and write commands take a parameter P.
Normally P is one byte (0-255).  If the command is preceded by
the Escape command then P is two bytes (0-65535, least significant
byte first).

.br
The address and flags default to 0.  The address and flags maintain
their previous value until updated.

.br
No flags are currently defined.

.br

\fBExample\fP
.br

.EX
Set address 0x53
.br
start, write 0x32, (re)start, read 6 bytes, stop
.br
Set address 0x1E
.br
start, write 0x03, (re)start, read 6 bytes, stop
.br
Set address 0x68
.br
start, write 0x1B, (re)start, read 8 bytes, stop
.br
End
.br

.br
0x04 0x53
.br
0x02 0x07 0x01 0x32   0x02 0x06 0x06 0x03
.br

.br
0x04 0x1E
.br
0x02 0x07 0x01 0x03   0x02 0x06 0x06 0x03
.br

.br
0x04 0x68
.br
0x02 0x07 0x01 0x1B   0x02 0x06 0x08 0x03
.br

.br
0x00
.br

.EE

.br

.IP "\fBBR1 \fP - Read bank 1 GPIO"
.IP "" 4
This command read GPIO 0-31 (bank 1) and returns the levels as a
32-bit hexadecimal value.

.br

\fBExample\fP
.br

.EX
$ pigs br1
.br
1001C1CF
.br

.EE

.br

.IP "\fBBR2 \fP - Read bank 2 GPIO"
.IP "" 4
This command read GPIO 32-53 (bank 2) and returns the levels as a
32-bit hexadecimal value.

.br

\fBExample\fP
.br

.EX
$ pigs br2
.br
003F0000
.br

.EE

.br

.IP "\fBBS1 bits\fP - Set specified GPIO in bank 1"
.IP "" 4
This command sets (sets high) the GPIO specified by \fBbits\fP in bank 1.
Bank 1 consists of GPIO 0-31.

.br
Upon success nothing is returned.  On error a negative status code
will be returned.

.br

\fBExample\fP
.br

.EX
$ pigs bs1 16 # set GPIO 4 (1<<4)
.br

.br
$ pigs bs1 1 # set GPIO 1 (1<<0)
.br
-42
.br
ERROR: no permission to update one or more GPIO
.br

.EE

.br

.IP "\fBBS2 bits\fP - Set specified GPIO in bank 2"
.IP "" 4
This command sets (sets high) the GPIO specified by \fBbits\fP in bank 2.
Bank 2 consists of GPIO 32-53.

.br
Upon success nothing is returned.  On error a negative status code
will be returned.

.br

\fBExample\fP
.br

.EX
$ pigs bs2 0x40 # set GPIO 38 (enable high current mode A+/B+/Pi2/Pi3)
.br

.br
$ pigs bs2 1 # set GPIO 32 (first in bank 2)
.br
-42
.br
ERROR: no permission to update one or more GPIO
.br

.EE

.br

.IP "\fBBSCX bctl bvs\fP - BSC I2C/SPI transfer"
.IP "" 4

.br
This command performs a BSC I2C/SPI slave transfer as defined by
\fBbctl\fP with data \fBbvs\fP.

.br
This function provides a low-level interface to the SPI/I2C Slave
peripheral on the BCM chip.

.br
This peripheral allows the Pi to act as a hardware slave device
on an I2C or SPI bus.

.br
This is not a bit bang version and as such is OS timing
independent. The bus timing is handled directly by the chip.

.br
The output process is simple. You simply append data to the FIFO
buffer on the chip.  This works like a queue, you add data to the
queue and the master removes it.

.br
The command sets the BSC mode and writes any data \fBbvs\fP
to the BSC transmit FIFO.  It returns the data count (at least 1
for the status word), the status word, followed by any data bytes
read from the BSC receive FIFO.

.br
Note that the control word sets the BSC mode.  The BSC will stay in
that mode until a different control word is sent.

.br
For I2C use a control word of (I2C address << 16) + 0x305.

.br
E.g. to talk as I2C slave with address 0x13 use 0x130305.

.br
GPIO used for models other than those based on the BCM2711.

.br

.EX
      SDA   SCL   MOSI   SCLK   MISO   CE
I2C   18    19    -      -      -      -
SPI   -     -     20     19     18     21

.EE

.br
GPIO used for models based on the BCM2711 (e.g. the Pi4B).

.br

.EX
      SDA   SCL   MOSI   SCLK   MISO   CE
I2C   10    11    -      -      -      -
SPI   -     -     9      11     10     8

.EE

.br
When a zero control word is received the used GPIO will be reset
to INPUT mode.

.br
The control word consists of the following bits.

.br

.EX
22 21 20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2  1  0
.br
 a  a  a  a  a  a  a  -  - IT HC TF IR RE TE BK EC ES PL PH I2 SP EN
.br

.EE

.br
Bits 0-13 are copied unchanged to the BSC CR register.  See
pages 163-165 of the Broadcom peripherals document for full
details.

.br

.EX
aaaaaaa   defines the I2C slave address (only relevant in I2C mode)
IT        invert transmit status flags
HC        enable host control
TF        enable test FIFO
IR        invert receive status flags
RE        enable receive
TE        enable transmit
BK        abort operation and clear FIFOs
EC        send control register as first I2C byte
ES        send status register as first I2C byte
PL        set SPI polarity high
PH        set SPI phase high
I2        enable I2C mode
SP        enable SPI mode
EN        enable BSC peripheral

.EE

.br
The returned status has the following format

.br

.EX
20 19 18 17 16 15 14 13 12 11 10  9  8  7  6  5  4  3  2  1  0
.br
 S  S  S  S  S  R  R  R  R  R  T  T  T  T  T RB TE RF TF RE TB
.br

.EE

.br
Bits 0-15 are copied unchanged from the BSC FR register.  See
pages 165-166 of the Broadcom peripherals document for full
details.

.br

.EX
SSSSS   number of bytes successfully copied to transmit FIFO
RRRRR   number of bytes in receieve FIFO
TTTTT   number of bytes in transmit FIFO
RB      receive busy
TE      transmit FIFO empty
RF      receive FIFO full
TF      transmit FIFO full
RE      receive FIFO empty
TB      transmit busy

.EE

.br
This example assumes that GPIO 2/3 are connected to GPIO 18/19
(GPIO 10/11 on the BCM2711).

.br

\fBExample\fP
.br

.EX
$ pigs bscx 0x130305 # start BSC as I2C slave 0x13
.br
1 18
.br

.br
$ i2cdetect -y 1
.br
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
.br
00:          -- -- -- -- -- -- -- -- -- -- -- -- --
.br
10: -- -- -- 13 -- -- -- -- -- -- -- -- -- -- -- --
.br
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
.br
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
.br
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
.br
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
.br
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
.br
70: -- -- -- -- -- -- -- --
.br

.br
$ pigs i2co 1 0x13 0 # get handle for device 0x13 on bus 1
.br
0
.br

.br
$ pigs i2cwd 0 90 87 51 9 23 # write 5 bytes
.br

.br
$ pigs bscx 0x130305 # check for data
.br
6 18 90 87 51 9 23
.br

.br
$ pigs bscx 0x130305 11 13 15 17 # check for data and send 4 bytes
.br
1 262338
.br

.br
$ pigs i2crd 0 4 # read 4 bytes
.br
4 11 13 15 17
.br

.br
$ pigs i2cwd 0 90 87 51 9 23 # write 5 bytes
.br
$ pigs bscx 0x130305 11 13 15 17 # check for data and send 4 bytes
.br
6 262338 90 87 51 9 23
.br

.br
$ pigs i2crd 0 4
.br
4 11 13 15 17
.br

.br
$ pigs bscx 0x130305 22 33 44 55 66
.br
1 327938
.br
$ pigs i2crd 0 5
.br
5 22 33 44 55 66
.br

.EE

.br
The BSC slave in SPI mode deserializes data from the MOSI pin into its receiver/
FIFO when the LSB of the first byte is a 0.  No data is output on the MISO pin.
When the LSB of the first byte on MOSI is a 1, the transmitter/FIFO data is
serialized onto the MISO pin while all other data on the MOSI pin is ignored.

.br
The BK bit of the BSC control register is non-functional when in the SPI mode.
The transmitter along with its FIFO can be dequeued by successively disabling
and re-enabling the TE bit on the BSC control register while in SPI mode.

.br
This example demonstrates a SPI master talking to the BSC as SPI slave:
Requires SPI master SCLK / MOSI / MISO / CE GPIO are connected to
BSC peripheral GPIO 11   / 9    / 10   / 8 respectively, on a Pi4B (BCM2711).

.br

\fBExample\fP
.br

.EX
$ pigs bspio 15 26 13 14 10000 0 # open bit-bang spi master on random gpio
.br

.br
$ pigs bscx 0x303 # start BSC as SPI slave, both rx and tx enabled
.br
1 18
.br

.br
$ pigs bspix 15 0 0xd 0xe 0xa 0xd # write 0xdead to BSC
.br
5 0 0 0 0 0
.br

.br
$ pigs bscx 0x303 0xb 0xe 0xe 0xf # place 0xbeef in BSC tx FIFO, read rx FIFO
.br
5 262338 13 14 10 13
.br

.br
$ pigs bspix 15 1 0 0 0 0 # read four bytes from BSC
.br
5 0 11 14 14 15
.br

.EE

.br

.IP "\fBBSPIC cs\fP - Close bit bang SPI"
.IP "" 4

.br
This command stops bit banging SPI on a set of GPIO
opened with \fBBSPIO\fP.

.br
The set of GPIO is specifed by \fBcs\fP.

.br
Upon success nothing is returned.  On error a negative status code
will be returned.

.br

\fBExample\fP
.br

.EX
$ pigs bspic 10
.br

.br
$ pigs bspic 10
.br
-142
.br

pigs.c
/*
This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org/>
*/

/*
This version is for pigpio version 69+
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <ctype.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netdb.h>
#include <arpa/inet.h>

#include "pigpio.h"
#include "command.h"
#include "pigs.h"

/*
This program provides a socket interface to some of
the commands available from pigpio.
*/

char command_buf[CMD_MAX_EXTENSION];
char response_buf[CMD_MAX_EXTENSION];

int printFlags = 0;

int status = PIGS_OK;

#define SOCKET_OPEN_FAILED -1

#define PRINT_HEX 1
#define PRINT_ASCII 2

void report(int err, char *fmt, ...)
{
   char buf[128];
   va_list ap;

   if (err > status) status = err;

   va_start(ap, fmt);
   vsnprintf(buf, sizeof(buf), fmt, ap);
   va_end(ap);

   fprintf(stderr, "%s\n", buf);

   fflush(stderr);
}

static int initOpts(int argc, char *argv[])
{
   int opt, args;

   opterr = 0;

   args = 1;

   while ((opt = getopt(argc, argv, "ax")) != -1)
   {
      switch (opt)
      {
         case 'a':
            printFlags |= PRINT_ASCII;
            args++;
            break;

         case 'x':
            printFlags |= PRINT_HEX;
            args++;
            break;

         default:
            args++;
            report(PIGS_OPTION_ERR, "ERROR: bad option %c", optopt);
      }
    }
   return args;
}

static int openSocket(void)
{
   int sock, err;
   struct addrinfo hints, *res, *rp;
   const char *addrStr, *portStr;

   portStr = getenv(PI_ENVPORT);

   if (!portStr) portStr = PI_DEFAULT_SOCKET_PORT_STR;

   addrStr = getenv(PI_ENVADDR);

   if (!addrStr) addrStr = PI_DEFAULT_SOCKET_ADDR_STR;

   memset (&hints, 0, sizeof (hints));

   hints.ai_family   = PF_UNSPEC;
   hints.ai_socktype = SOCK_STREAM;
   hints.ai_flags   |= AI_CANONNAME;

   err = getaddrinfo(addrStr, portStr, &hints, &res);

   if (err) return SOCKET_OPEN_FAILED;

   for (rp=res; rp!=NULL; rp=rp->ai_next)
   {
      sock = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);

      if (sock == -1) continue;

      if (connect(sock, rp->ai_addr, rp->ai_addrlen) != -1) break;
   }

   freeaddrinfo(res);

   if (rp == NULL) return SOCKET_OPEN_FAILED;

   return sock;
}

void print_result(int sock, int rv, cmdCmd_t cmd)
{
   int i, r, ch;
   uint32_t *p;

   r = cmd.res;

   switch (rv)
   {
      case 0:
      case 1:
         if (r < 0)
         {
            printf("%d\n", r);
            report(PIGS_SCRIPT_ERR, "ERROR: %s", cmdErrStr(r));
         }
         break;

      case 2:
         printf("%d\n", r);
         if (r < 0) report(PIGS_SCRIPT_ERR, "ERROR: %s", cmdErrStr(r));
         break;

      case 3:
         printf("%08X\n", cmd.res);
         break;

      case 4:
         printf("%u\n", cmd.res);
         break;

      case 5:
         printf("%s", cmdUsage);
         break;

      case 6: /*
                 BI2CZ  CF2  FL  FR  I2CPK  I2CRD  I2CRI  I2CRK
                 I2CZ  SERR  SLR  SPIX  SPIR
              */
         printf("%d", r);
         if (r < 0) report(PIGS_SCRIPT_ERR, "ERROR: %s", cmdErrStr(r));
         if (r > 0)
         {
            if (printFlags == PRINT_ASCII) printf(" ");

            for (i=0; i<r; i++)
            {
               ch = response_buf[i];

               if (printFlags & PRINT_HEX) printf(" %x", ch);

               else if (printFlags & PRINT_ASCII)
               {
                  if (isprint(ch) || (ch == '\n') || (ch == '\r'))
                     printf("%c", ch);
                  else printf("\\x%02x", ch);
               }
               else printf(" %hhu", response_buf[i]);
            }
         }
         printf("\n");
         break;

      case 7: /* PROCP */
         if (r != (4 + (4*PI_MAX_SCRIPT_PARAMS)))
         {
            printf("%d", r);
            report(PIGS_SCRIPT_ERR, "ERROR: %s", cmdErrStr(r));
         }
         else
         {
            p = (uint32_t *)response_buf;
            printf("%d", p[0]);
            for (i=0; i<PI_MAX_SCRIPT_PARAMS; i++)
            {
               printf(" %d", p[i+1]);
            }
         }
         printf("\n");
         break;

      case 8: /*
                 BSCX
              */
         if (r < 0)
         {
            printf("%d\n", r);
            report(PIGS_SCRIPT_ERR, "ERROR: %s", cmdErrStr(r));
            break;
         }

         p = (uint32_t *)response_buf;
         printf("%d %d", r-3, p[0]);

         if (r > 4)
         {
            if (printFlags == PRINT_ASCII) printf(" ");

            for (i=4; i<r; i++)
            {
               ch = response_buf[i];

               if (printFlags & PRINT_HEX) printf(" %x", ch);

               else if (printFlags & PRINT_ASCII)
               {
                  if (isprint(ch) || (ch == '\n') || (ch == '\r'))
                     printf("%c", ch);
                  else printf("\\x%02x", ch);
               }
               else printf(" %hhu", response_buf[i]);
            }
         }
         printf("\n");
         break;

   }
}

void get_extensions(int sock, int command, int res)
{
   switch (command)
   {
      case PI_CMD_BI2CZ:
      case PI_CMD_BSCX:
      case PI_CMD_BSPIX:
      case PI_CMD_CF2:
      case PI_CMD_FL:
      case PI_CMD_FR:
      case PI_CMD_I2CPK:
      case PI_CMD_I2CRD:
      case PI_CMD_I2CRI:
      case PI_CMD_I2CRK:
      case PI_CMD_I2CZ:
      case PI_CMD_PROCP:
      case PI_CMD_SERR:
      case PI_CMD_SLR:
      case PI_CMD_SPIX:
      case PI_CMD_SPIR:

         if (res > 0)
         {
            recv(sock, response_buf, res, MSG_WAITALL);
            response_buf[res] = 0;
         }
         break;
   }
}

int main(int argc , char *argv[])
{
   int sock, command;
   int args, idx, i, pp, l, len;
   cmdCmd_t cmd;
   uintptr_t p[CMD_P_ARR];
   cmdCtlParse_t ctl;
   cmdScript_t s;
   char v[CMD_MAX_EXTENSION];

   sock = openSocket();

   args = initOpts(argc, argv);

   command_buf[0] = 0;
   l = 0;
   pp = 0;

   for (i=args; i<argc; i++)
   {
      l += (strlen(argv[i]) + 1);
      if (l < sizeof(command_buf))
         {sprintf(command_buf+pp, "%s ", argv[i]); pp=l;}
   }

   if (pp) {command_buf[--pp] = 0;}

   ctl.eaten = 0;

   len = strlen(command_buf);
   idx = 0;

   while ((idx >= 0) && (ctl.eaten < len))
   {
      if ((idx=cmdParse(command_buf, p, CMD_MAX_EXTENSION, v, &ctl)) >= 0)
      {
         command = p[0];

         if (command < PI_CMD_SCRIPT)
         {
            if (command == PI_CMD_HELP)
            {
               printf("%s", cmdUsage);
            }
            else if (command == PI_CMD_PARSE)
            {
               cmdParseScript(v, &s, 1);
               if (s.par) free (s.par);
            }
            else
            {
               cmd.cmd = command;
               cmd.p1 = p[1];
               cmd.p2 = p[2];
               cmd.p3 = p[3];

               if (sock != SOCKET_OPEN_FAILED)
               {
                  if (send(sock, &cmd, sizeof(cmdCmd_t), 0) ==
                     sizeof(cmdCmd_t))
                  {
                     if (p[3]) send(sock, v, p[3], 0); /* send extensions */

                     if (recv(sock, &cmd, sizeof(cmdCmd_t), MSG_WAITALL) ==
                        sizeof(cmdCmd_t))
                     {
                        get_extensions(sock, command, cmd.res);

                        print_result(sock, cmdInfo[idx].rv, cmd);
                     }
                     else report(PIGS_CONNECT_ERR, "socket receive failed");
                  }
                  else report(PIGS_CONNECT_ERR, "socket send failed");
               }
               else report(PIGS_CONNECT_ERR, "socket connect failed");
            }
         }
         else report(PIGS_SCRIPT_ERR,
                 "%s only allowed within a script", cmdInfo[idx].name);
      }
      else
      {
         if (idx == CMD_UNKNOWN_CMD)
            report(PIGS_SCRIPT_ERR,
               "%s? unknown command, pigs h for help", cmdStr());
         else
            report(PIGS_SCRIPT_ERR,
               "%s: bad parameter, pigs h for help", cmdStr());
      }
   }

   if (sock >= 0) close(sock);

   return status;
}

